### PR TITLE
[SPARK-34571][SQL][CORE] Provide a more convenient way to deprecate/remove/alternate configs

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This pr proposes to add `deprecated()`/ `removed()` / `alternated()` functions to `ConfigBuilder` in order to provide more convenient ways to do these stuff.

### Why are the changes needed?

Currently, Spark maintains the active configs and deprecated/alternated/removed configs separately,
which is not convenient when developers want to deprecate configs and not good to maintain for the long term.

After this PR, developers can deprecate/alternate/remove a config like this:

```scala
val SOME_CONFIG =
  buildConf("spark.xxx.yyy.zzz")
    .doc(".....")")
    .version("2.3.0")
    .deprecated("3.0", "because...")
    // .removed("3.0", "because...")
    // .alternated(("3.0", "spark.aaa.bbb.ccc"))
    .booleanConf
    .createWithDefault(false)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually tested.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan @HyukjinKwon @MaxGekk any idea about this?

**Review by MaxGekk — COMMENTED** [review]
> This pr proposes to add deprecated()/ removed() / alternated() functions to ConfigBuilder in order to provide more convenient ways to do these stuff.

I think if you propose new approach, you should restore all removed configs and make them as `.removed()` then. Don't think it is good idea to maintain two approach at the same time.

> maintains the active configs and deprecated/alternated/removed configs separately,
which is not convenient when developers want to deprecate configs and not good to maintain for the long term.

IMHO, gathering all removed/deprecated configs in one place is convenient. Currently, removed configs are actually removed but you propose to keep them forever (just mark them as removed), am I right?

**Comment by Ngone51** [comment]
> I think if you propose new approach, you should restore all removed configs and make them as .removed() then. Don't think it is good idea to maintain two approach at the same time.

Yeah, I was thinking that too. But I'd like to see people's feedback before going further.

> IMHO, gathering all removed/deprecated configs in one place is convenient. 

I think it's convenient (fast) for people to know what are deprecated or removed configs now. But it's not convenient for developers to maintain. For example, to deprecate a config with the current way, a developer needs to manually add `DepracatedConfig` to the deprecated list. And if the config is going to be removed in the future, the developer has to remove the `DepracatedConfig` from the deprecated list and add `RemovedConfig` to the removed list. And this process might be problematic because these two config types are separated in two places. e.g., developers may forget to remove the `DepracatedConfig` one. While with the proposed way in this PR, the developer only needs to add `deprecated()` and change from `deprecated()` to `removed()` later.

> Currently, removed configs are actually removed but you propose to keep them forever (just mark them as removed), am I right?

Yes. But those removed configs don't disappear from Spark totally as they are still maintained in the way of `RemovedConfig`.

**Comment by Ngone51** [comment]
Besides, I'm thinking we may be able to extend the SQL configuration document with the new proposal. For example, we can add the current status of a conf, e.g., whether it's deprecated or removed, etc. like the `version` info does. cc @beliefer

File: core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala [hunk 1]
```diff
@@ -205,6 +206,23 @@ private[spark] case class ConfigBuilder(key: String) {
     this
   }
 
+  def deprecated(version: String, deprecationMessage: String): ConfigBuilder = {
+    LegacyConfigsRegister.registerDeprecated(key, version, deprecationMessage)
+    this
+  }
+
+  def alternated(
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala:214]: nit but `alternated` -> `alternative`

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 2]
```diff
@@ -808,9 +832,21 @@ private[spark] object SparkConf extends Logging {
    * @param version The Spark version in which the key was deprecated.
    * @param translation A translation function for converting old config values into new ones.
    */
-  private case class AlternateConfig(
+  private[spark] case class AlternateConfig(
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkConf.scala:835]: Let's fix all Alternate -> Alternative

**Comment by HyukjinKwon** [comment]
Can we add deprecate and alternative ones first? For removal one, I would encourage to do it in a separate PR.

**Comment by cloud-fan** [comment]
Agree with @HyukjinKwon to do deprecation first. For the removed ones, we don't really need to keep the config entries for them anymore. We just need the names of removed configs (and the successors) to throw better error messages.

**Comment by beliefer** [comment]
@Ngone51 Record the version when deciding deprecate config, seems good.

**Comment by Ngone51** [comment]
Thanks for all the feedbacks..I'll start from the depreaction and alternatives first.

File: core/src/main/scala/org/apache/spark/SparkConf.scala [hunk 3]
```diff
@@ -613,31 +628,7 @@ private[spark] object SparkConf extends Logging {
       DeprecatedConfig("spark.yarn.services", "3.0.0", "Feature no longer available."),
       DeprecatedConfig("spark.executor.plugins", "3.0.0",
         "Feature replaced with new plugin API. See Monitoring documentation."),
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkConf.scala:630]:
```suggestion
        "Feature replaced with new plugin API. See Monitoring documentation.")
```

**Pull request closed without merging**